### PR TITLE
THEEDGE-3019 Fixes missing data when a device update finishes

### PR DESCRIPTION
**Pull request by de1987**
# Description
**AvailableHash** and **CurrentHash** are not being saved when the device update finishes.

FIXES: [THEEDGE-3019](https://issues.redhat.com/browse/THEEDGE-3019)

## Type of change

What is it?

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [x] Tests update
- [ ] Refactor

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I run `make pre-commit` to check fmt/vet/lint/test-no-fdo

### CODE REVIEW

**Review by omaciel — APPROVED** [review]
/lgtm

FWIW I feel that we need more testing around anything we're calling with `Omit` for database transactions. I don't yet understand why we include both `Device` **and** `DeviceID` as part of pkg.models.updates.go#DispatchRecord:


```go
type DispatchRecord struct {
	Model
	PlaybookURL          string  `json:"PlaybookURL"`
	DeviceID             uint    `json:"DeviceID"`
	Device               *Device `json:"Device"`
	Status               string  `json:"Status"`
	Reason               string  `json:"Reason"`
	PlaybookDispatcherID string  `json:"PlaybookDispatcherID"`
}
```

File: pkg/services/updates.go [hunk 1]
```diff
@@ -561,6 +561,12 @@ func (s *UpdateService) ProcessPlaybookDispatcherRunEvent(message []byte) error
 		return result.Error
 	}
 
+	// since it's using Omit, the device is not being saved, then it's required to explicit save the device
+	result = db.DB.Save(&dispatchRecord.Device)
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply pkg/services/updates.go:565]: ah Ok I see you added the device check in the unit-tests, for unittest part it should be sufficient.

File: pkg/services/updates.go [hunk 2]
```diff
@@ -561,6 +561,12 @@ func (s *UpdateService) ProcessPlaybookDispatcherRunEvent(message []byte) error
...
+	if result.Error != nil {
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply pkg/services/updates.go:566]: for Device and DeviceID in modes.DispatchRecord, this how belong-to relationships works in gorm https://gorm.io/docs/belongs_to.html

[the remark this answers is not in the document]
Comment (omaciel) [reply pkg/services/updates.go:566]: Thanks @ldjebran

**Comment by de1987** [comment]
> /lgtm
> 
> FWIW I feel that we need more testing around anything we're calling with `Omit` for database transactions. I don't yet understand why we include both `Device` **and** `DeviceID` as part of pkg.models.updates.go#DispatchRecord:
> 
> ```go
> type DispatchRecord struct {
> 	Model
> 	PlaybookURL          string  `json:"PlaybookURL"`
> 	DeviceID             uint    `json:"DeviceID"`
> 	Device               *Device `json:"Device"`
> 	Status               string  `json:"Status"`
> 	Reason               string  `json:"Reason"`
> 	PlaybookDispatcherID string  `json:"PlaybookDispatcherID"`
> }
> ```



> /lgtm
> 
> FWIW I feel that we need more testing around anything we're calling with `Omit` for database transactions. I don't yet understand why we include both `Device` **and** `DeviceID` as part of pkg.models.updates.go#DispatchRecord:
> 
> ```go
> type DispatchRecord struct {
> 	Model
> 	PlaybookURL          string  `json:"PlaybookURL"`
> 	DeviceID             uint    `json:"DeviceID"`
> 	Device               *Device `json:"Device"`
> 	Status               string  `json:"Status"`
> 	Reason               string  `json:"Reason"`
> 	PlaybookDispatcherID string  `json:"PlaybookDispatcherID"`
> }
> ```

Let's discuss it with the team, I agree with you.
Just being caution because this would require some refactor in the code.

File: pkg/services/updates.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (de1987) [reply pkg/services/updates.go:565]: hey Djebran, yes I added
before it was failing and this is why we fixed it and updated the tests

**Pull request merged**